src/macros.jl

```diff
@@ -184,9 +184,40 @@ function sense_to_set(_error::Function, ::Val{S}) where {S}
     return _error("Unrecognized sense $S")
 end
 
+"""
+    _desparsify(x)
+
+If `x` is an `AbstractSparseArray`, return the dense equivalent, otherwise just
+return `x`.
+
+## Why is this needed?
+
+When broadcasting `f.(x)` over an `AbstractSparseArray` `x`, Julia first calls
+the equivalent of `f(zero(eltype(x))`. However, if `f` is mutating, this can
+have serious consequences! In our case, broadcasting `build_constraint` will add
+a new `0 = 0` constraint.
```

Comment:
How does `collect` help ? It also replaces the structural zero elements with zeros

Reply:
In most cases (all cases?), promotion means that the incoming SparseArray is actually dense.

In 0.21.7, you get
```Julia
julia> model = Model()
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: NO_OPTIMIZER
Solver name: No optimizer attached.

julia> @variable(model, x[1:2])
2-element Vector{VariableRef}:
 x[1]
 x[2]

julia> d = sparsevec([1, 3], x)
3-element SparseVector{VariableRef, Int64} with 2 stored entries:
  [1]  =  x[1]
  [3]  =  x[2]

julia> @constraint(model, d .== 0)
3-element SparseVector{Any, Int64} with 3 stored entries:
  [1]  =  x[1] = 0.0
  [2]  =  0 = 0.0
  [3]  =  x[2] = 0.0

julia> model
A JuMP Model
Feasibility problem with:
Variables: 2
`AffExpr`-in-`MathOptInterface.EqualTo{Float64}`: 4 constraints
Model mode: AUTOMATIC
CachingOptimizer state: NO_OPTIMIZER
Solver name: No optimizer attached.
Names registered in the model: x
```

In this PR, you get
```Julia
julia> model = Model()
A JuMP Model
Feasibility problem with:
Variables: 0
Model mode: AUTOMATIC
CachingOptimizer state: NO_OPTIMIZER
Solver name: No optimizer attached.

julia> @variable(model, x[1:2])
2-element Vector{VariableRef}:
 x[1]
 x[2]

julia> d = sparsevec([1, 3], x)
3-element SparseVector{VariableRef, Int64} with 2 stored entries:
  [1]  =  x[1]
  [3]  =  x[2]

julia> @constraint(model, d .== 0)
3-element Vector{ConstraintRef{Model, MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64}, MathOptInterface.EqualTo{Float64}}, ScalarShape}}:
 x[1] = 0.0
 0 = 0.0
 x[2] = 0.0

julia> model
A JuMP Model
Feasibility problem with:
Variables: 2
`AffExpr`-in-`MathOptInterface.EqualTo{Float64}`: 3 constraints
Model mode: AUTOMATIC
CachingOptimizer state: NO_OPTIMIZER
Solver name: No optimizer attached.
Names registered in the model: x
```

Is changing the return type of `@constraint` here breaking? The type change happens with dense constraints too.